{{ cookiecutter.repo_name }}/.github/workflows/ci.yml

```diff
@@ -8,18 +8,7 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v1
-      - name: Black
-        uses: ./docker/action
-        with:
-          command: black --check {{ cookiecutter.repo_name }}
-      - name: Flake8
-        uses: ./docker/action
-        with:
-          command: flake8 {{ cookiecutter.repo_name }}
-      - name: Unit Tests
-        uses: ./docker/action
-        with:
-          command: pytest {{ cookiecutter.repo_name }}
-        env:
-          MLFLOW_TRACKING_URI: {{ cookiecutter.mlflow_uri }}
-          MLFLOW_ARTIFACT_LOCATION: {{ cookiecutter.mlflow_artifact }}
+      - name: Build Docker container
+        run: touch .env && make dev-start
```

Comment:
This seems amazing. But i'm a little unclear on how this magic works? On the GitHub Actions server we are spinning up a Docker container and then running the CI inside it. And it just works? Do we lose any Docker layer caching facility by doing this? Said another way -- does using the Docker GitHub action have some benefits that we lose by "rolling our own" in this sense?

Reply:
TLDR, we don't lose anything from what is being used currently---both options are effectively "rolling our own" solution, and they do the same thing.

What is happening on the GitHub Actions server is the same as what happens locally when we call `make dev-start`: we build the Docker container on the server using the project's Dockerfile, then run the CI inside the container. The image is cached on the server for the duration of the job. So each time we run the CI between `git push`'s, the server has to rebuild the layers beyond the base image.

The only reason I built the Docker action was because the old workflow file just ran `black`, etc. in a container built from the project's base Docker image. I wanted to run those same commands in a container built from the project Dockerfile, but I couldn't find a straight-forward way to do that, so I read the GitHub Workflow/Aorkflow docs and learned that we could only do that by making a custom action. The Docker GitHub action I made was basically just a wrapper for `docker run <project_container> <command>` using the container built from the project Dockerfile. It also had to rebuild the layers beyond the base image and was only cached for the duration of the job.

But the new workflow file does the same thing, from the server's point-of-view: build the project's Docker image from the Dockerfile, spin up a container, and run the CI inside the container.

Reply:
Makes sense. Carry on then :)